### DISCUSSION TITLE
RE: design for cachi2 wheels: Report whether dependency has any wheels (in the SBOM)

### DISCUSSION THREAD

**Discussion by ben-alkov** [containerbuildsystem/cachi2#451] [Q&A] [answered]
@brunoapimentel, @ejegrova, @eskultety, @slimreaper35, @taylormadore;

Is https://github.com/containerbuildsystem/cachi2/blob/6225d7a03928951ee799f88710abd6dd58f78b34/cachi2/core/package_managers/pip.py#L122-L126 the right place to implement adding a property to an SBOM component (for Python wheels)?

> Note to self: Wheels were implemented in https://github.com/containerbuildsystem/cachi2/pull/255 & https://github.com/containerbuildsystem/cachi2/pull/319

**Comment by eskultety** [comment]
IIUC yes, it is.

**Comment by ben-alkov** [comment]
> IIUC yes, it is

Nice. OK, second question - what determines whether a dependency is a wheel or not? I've tried

- `dependency["package_type"] == "wheel"`
- hacking in (what I thought was) `dependency["wheel"]`

but neither of those properties exist by the time the dependecy shows up in `_generate_properties()`...

**Reply by ben-alkov** [reply]
It looks like the `allow_binary` flag is actually (finally) used in
https://github.com/containerbuildsystem/cachi2/blob/6225d7a03928951ee799f88710abd6dd58f78b34/cachi2/core/package_managers/pip.py#L1459-L1472

but the careful observer will note that `to_download` is never appended to `downloaded`, which is the *only* return from `_download_dependencies()`

Also, I note with interest that none of the docstrings have been updated to include `allow_binary`, which is making this all the more difficult.

**Reply by ben-alkov** [reply]
Oh, and another thing; every place in the cachi2 source where we are _assigning_ to `allow_binary`, it is set to `False`. The design doc *explicitly* says to do the opposite?![1]

[1] Section **Option not to download wheels**: "The default should be allow_binary: true, since that usually results in less surprising behavior"

**Reply by brunoapimentel** [reply]
As for the `allow_binary` defaulting to False for now, I think it's because we didn't actually release the feature yet. Once it's done and documented, we can turn it on to `true` by default, but we also need to think on the impact this would have on Konflux (as it is the main Cachi2 'user' nowadays).

**Reply by chmeliik** [reply]
It probably shouldn't default to true until Enterprise Contract can block builds that use wheels (if it should ever default to true)

**Comment by ben-alkov** [comment]
I'm still blocked on *exactly* what marks a dependency as a wheel, **permanently** (i.e. after all downloading etc. is complete)... as I noted above, `to_download` is never appended to `downloaded`, which is the *only* return from `_download_dependencies()`.

**Reply by slimreaper35** [reply]
I think there is another story that should fix this "problem" and make handling all Python packages more straightforward.

https://issues.redhat.com/browse/STONEBLD-1705

@chmeliik Am I right?

**Reply by chmeliik** [reply, marked as answer]
Hmm, yeah Ben's story would probably (require / be easier to do after) this refactoring

**Reply by ben-alkov** [reply]
> Hmm, yeah Ben's story would probably (require / be easier to do after) this refactoring

Phew, ok, at least it's not just me :roll_eyes:. I was starting to think my brains had turned to oatmeal.

**Comment by ben-alkov** [comment]
Resolved in #467 and #508
